executor/src/witgen/range_constraints.rs

```diff
@@ -10,77 +12,202 @@ use super::symbolic_evaluator::SymbolicEvaluator;
-    pub fn try_combine_sum(&self, other: &RangeConstraint<T>) -> Option<RangeConstraint<T>> {
-        if self.mask & other.mask == 0u32.into() {
-            Some(RangeConstraint {
-                mask: self.mask | other.mask,
-            })
+    pub fn combine_sum(&self, other: &Self) -> Self {
+        let mask = (self.mask + other.mask) | self.mask | other.mask;
+
+        // TODO test the case where it is equal to modulus
+        let (min, max) = if self.range_width().to_arbitrary_integer()
+            + other.range_width().to_arbitrary_integer()
+            <= T::modulus().to_arbitrary_integer()
+        {
+            (self.min + other.min, self.max + other.max)
         } else {
-            None
-        }
+            (T::one(), T::zero())
+        };
+        Self { min, max, mask }
     }
 
     /// Returns the conjunction of this constraint and the other.
-    pub fn conjunction(self, other: &RangeConstraint<T>) -> RangeConstraint<T> {
-        RangeConstraint {
-            mask: self.mask & other.mask,
+    pub fn conjunction(&self, other: &Self) -> Self {
+        let mut mask = self.mask & other.mask;
+        // TODO this should be correct, we need a proof.
+        let mut min = cmp::max(self.min, other.min);
+        let mut max = cmp::min(self.max, other.max);
+
+        // Now try to derive better values for the mask from the new range
+        // and vice-versa.
+        if mask < T::modulus() {
+            if min <= max {
+                // If we adjust both min and max, the right way could be
+                // to have an empty range. On the other hand, this should not
+                // be incorrect.
+                min = cmp::min(mask.into(), min);
+                max = cmp::min(mask.into(), max);
+            } else if min.to_integer() > mask {
+                min = T::zero();
+                max = cmp::min(mask.into(), max);
+            } else {
+                // max < min <= mask
+                // TODO
+            }
         }
+        if min <= max {
+            mask &= mask_from_bits::<T>(max.to_integer().num_bits() as u64);
+        }
+
+        Self { min, max, mask }
     }
 
-    /// The bit constraint of an integer multiple of an expression.
-    pub fn multiple(&self, factor: T) -> Option<RangeConstraint<T>> {
-        let exponent = log2_exact(factor.to_arbitrary_integer())?;
-        let mask = self.mask << exponent;
-        if mask.to_arbitrary_integer() < T::modulus().to_arbitrary_integer() {
-            Some(RangeConstraint { mask })
+    /// The constraint of an integer multiple of an expression.
+    pub fn multiple(&self, factor: T) -> Self {
+        // TODO If factor is 1 or -1, we can also work with range constraints.
+        let mask = log2_exact(factor.to_arbitrary_integer()).and_then(|exponent| {
+            (self.mask.to_arbitrary_integer() << exponent < T::modulus().to_arbitrary_integer())
+                .then_some(self.mask << exponent)
```

Comment:
If `then_some` is not inlined or optimized, the inner expression will have be calculated (contrary to using `if`).